streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,638 @@
+            decompressing downloaded shards. If set to None, keep iff remote is local. Defaults to
+            ``None``.
+        download_retry (int): Number of download re-attempts before giving up. Defaults to ``2``.
+        download_timeout (float): Number of seconds to wait for a shard to download before raising
+            an exception. Defaults to ``60``.
+        validate_hash (str, optional): Optional hash or checksum algorithm to use to validate
+            shards. Defaults to ``None``.
+        seed (int, optional): Seed for shuffling, or ``None`` for random seed. Defaults to
+            ``None``.
+        batch_size (int, optional): Batch size of its DataLoader, which affects how the dataset is
+            partitioned over the workers. Defaults to ``None``.
+        num_workers (int, optional): Number of workers of its DataLoader, which determines the size
+            of the barrier to coordinate workers while iterating. Defaults to ``None``.
+    """
+
+    def __init__(self,
+                 local: str,
+                 remote: Optional[str] = None,
+                 split: Optional[str] = None,
+                 shuffle: bool = False,
+                 predownload: Optional[int] = 100_000,
+                 keep_zip: Optional[bool] = None,
+                 download_retry: int = 2,
+                 download_timeout: float = 60,
+                 validate_hash: Optional[str] = None,
+                 seed: Optional[int] = None,
+                 batch_size: Optional[int] = None,
+                 num_workers: Optional[int] = None):
+        self.local = local
+        self.remote = remote
+        self.split = split or ''  # Empty string for os.path.join().
+        self.shuffle = shuffle
+        self.predownload = predownload
+        self.keep_zip = keep_zip
+        self.download_retry = download_retry
+        self.download_timeout = download_timeout
+        self.validate_hash = validate_hash or None
+        # Seed is set below.
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+
+        # Load the index.json file.
+        basename = get_index_basename()
+        world = World()
+        if world.is_local_leader:
+            filename = self._download_file(basename)
+        else:
+            filename = os.path.join(local, split, basename)  # pyright: ignore
+        dist.barrier()
+        obj = json.load(open(filename))
+        assert obj['version'] == 2
+
+        # Initialize shard readers according to the loaded info.
+        self.shards = []
+        for info in obj['shards']:
+            shard = reader_from_json(local, split, info)
+            self.shards.append(shard)
+
+        # Build the Index (for partitioning and mapping samples to shards).
+        self.shard_sizes = np.array([x.samples for x in self.shards])
+        self.index = Index(self.shard_sizes)
+
+        # Setup for coordinating.
```

Comment:
Can we move line 137 to 153 in a separate method? `__init__()` method is too crowded.